[Advanced settings] Add settings allowlist

ElenaStoeva opened the pull request:
Addresses https://github.com/elastic/kibana/issues/160411

## Summary

This PR adds functionality for filtering out advanced settings that are not relevant for serverless.

For context, we need to build an Advanced settings page in serverless which only contains a set of the existing settings. We will reuse the section registry (https://github.com/elastic/kibana/pull/163502) from the original Advanced settings plugin as well as its UI components which will also be extracted into a separate package. The app will be registered from inside the `serverless` plugin.

In order to only display the settings that are relevant for serverless, we need to make some changes to the uiSettings service. The implementation in this PR leverages the existing `readonly` uiSettings param and adds the `setAllowlist()` method which is called by the serverless plugin to set an allowlist of setting keys.

**Testing in serverless:**
1. Set `advanced_settings.enabled: true` to enable the Advanced settings app in serverless: https://github.com/elastic/kibana/blob/5b216c6ea94e739fea1f161f0bbce5a57ae44c02/config/serverless.yml#L53
2. Start Es with `yarn es serverless --ssl` and Kibana with `yarn serverless-{mode} --ssl` in any serverless mode.
3. Navigate to `app/management/kibana/settings`
4. Verify that the app only displays the settings from `packages/serverless/settings/common/index.ts` (these are the settings, relevant for all projects in serverless) as well as the settings from the corresponding project package `packages/serverless/settings/{mode}_project/index.ts`.
5. Verify that the app is functioning correctly.

**Testing in self-managed:**
1. Start Es with `yarn es snapshot` and Kibana with `yarn start`.
2. Go to Stack Management > Advanced settings
3. Verify that all settings are displayed as usual.
4. Verify that the app is functioning correctly.

If your team is a code owner of any of the serverless project plugins, please review the corresponding package `packages/serverless/settings/{search/observanility/security}_project/index.ts` where you've been added as an owner and test in the serverless solution accordingly.


<!---
### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |

-->

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

ElenaStoeva committed:
[Advanced settings] Add settings allowlist

lukeelmers commented on packages/core/ui-settings/core-ui-settings-common/src/ui_settings.ts:
```diff
@@ -92,6 +92,10 @@ export interface UiSettingsParams<T = unknown> {
    * scoped to a namespace. The default value is 'namespace'
    */
   scope?: UiSettingsScope;
+  /**
+   * Specifies whether the setting should be displayed in the Advanced settings app.
+   */
+  allowlisted?: boolean;
```
just a nit, we try to keep core unaware of other parts of kibana like advanced settings. I might consider renaming this to be something more generic like `readonly`

lukeelmers replied:
I forgot about the readonly param and looked into it further. It appears to hide the setting from the UI only, but doesn't seem to actually prevent changing the setting via the `/internal/kibana/settings` API.

Thinking aloud:

Maybe it would make more sense to leverage that existing flag, and add enforcement to it in the API (e.g. if a setting is `readonly: true`, you cannot set it). Then the API in core could basically allow manipulating this value, something like `coreSetup.uiSettings.setReadOnly(['*'])` and then `coreSetup.uiSettings.setWritable([...])`? Need to think about that some more, @clintandrewhall any opinions?

We'd need to consider whether it is safe to roll out enforcement to self-managed since it would affect any of the existing settings that are readonly, and customers may be relying on being able to change those programmatically via API. Using a separate flag like `frozen` or `enabled` would solve this problem, it's just sort of redundant I guess since `readonly` is the behavior we care about here.

There might not be a good way around having another flag, but long-term I'd suggest we push to only use `readonly` for this purpose.

clintandrewhall replied:
> Then the API in core could basically allow manipulating this value, something like `coreSetup.uiSettings.setReadOnly(['*'])` and then `coreSetup.uiSettings.setWritable([...])`?

@luke I think this makes it a bit confusing about what's been called where, and the status of the any/all of the settings.

Another option would be to allow an API on the service to define the _nature_ of readonly.  For example, something like `coreSetup.uiSettings.setReadOnlyState('hidden' | 'frozen')`.  Then the behavior in fully-managed doesn't change.  Setting an allow-list could toggle this, but at least we don't change the nature of `readonly`.

lukeelmers replied:
> I think this makes it a bit confusing about what's been called where, and the status of the any/all of the settings.

Yeah that's fair, I'm struggling to think of what the simplest API would be here. As we discussed offline, perhaps we stick with `setAllowlist` and as a side effect enforce it at the API level. Then in 9.0, we can change the behavior of `readonly` in classic so that we are always enforcing `readonly` in the APIs as well. This is where I'd like to get more 👀 from the core team.

lukeelmers commented on src/plugins/advanced_settings/public/management_app/settings_allowlist.ts:
```diff
@@ -0,0 +1,62 @@
+
+const SEARCH_SETTINGS = ['courier:ignoreFilterIfFieldNotInIndex', 'search:timeout'];
+
+export const ALL_ALLOWLIST_SETTINGS = [
```
the plan was to set this stuff from the `serverless` plugin, right?

ElenaStoeva replied:
That's correct. I'm only testing it in the original advanced settings plugin.

@clintandrewhall is currently working on extracting the Advanced settings UI components into a package, and once it is done, we will reuse it in the `serverless` plugin so that we register the Advanced settings app there. Then we can call `setAllowlist()` from the `serverless` plugin and put it all together.

lukeelmers reviewed (commented):
Thanks for pinging me on this! I added some initial thoughts on the overall approach. I think the core team will have more comments around the design & naming of the APIs, but if you started by moving everything to the server I think it would help make clear what you're trying to accomplish here, and ultimately make it easier for everyone else to review the proposal.

ElenaStoeva committed:
Move allowlist APIs to server

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

ElenaStoeva commented on packages/core/ui-settings/core-ui-settings-browser-internal/src/ui_settings_client_common.ts:
```diff
@@ -100,6 +100,22 @@ You can use \`IUiSettingsClient.get("${key}", defaultValue)\`, which will just r
+  allowlistAll() {
+    Object.entries(this.defaults).forEach(([key, value]) => (value.allowlisted = true));
+    Object.entries(this.cache).forEach(([key, value]) => (value.allowlisted = true));
+  }
```
> to have a new API along the lines of `coreSetup.uiSettings.setAllowlist([...]);` which can only be called once (or it throws an error).

Based on my understanding, we want to allowlist the settings for all serverless projects from the serverless plugin, the setttings for the security project from serverless security plugin, etc. (@clintandrewhall can you confirm that this is still the plan?). Therefore, we might need to call `setAllowlist()` two times (once from the serverless plugin and once from the serverless plugin for the specific project). Would there be an issue with this?

Alternatively, if we want to only allow calling it once, I suppose we could only call it from the serverless project plugins (i.e. the serverless security plugin allowlisting the general settings + the security-specific settings).

> Then we'd need to throw an error if someone tried to `set` or `get` a setting that wasn't allowed, or if they tried to do the same via the HTTP APIs as mentioned in my other comment.

I'm wondering if this might not break something since there might be settings that we don't want to display in the Advanced settings app, but are consumed by other plugins in Kibana. For example, I see that there is a setting [theme:version](https://github.com/elastic/kibana/blob/5aee5da843fc1e7747b66bb45e8ebf7a808c144d/packages/core/ui-settings/core-ui-settings-server-internal/src/settings/theme.ts#L60) which has the `readonly` flag set to true and so it's not displayed in the UI but is accessed by some places. If we don't allow getting settings that are not allowlisted, this setting would also not be accessible. Probably the same applies for settings with the `isCustom` flag as they are also [hidden in the UI](https://github.com/elastic/kibana/blob/5aee5da843fc1e7747b66bb45e8ebf7a808c144d/src/plugins/advanced_settings/public/management_app/settings_helper.ts#L27).

Overall, I thought it would be okay for other places in Kibana to `get` and `set` settings that are not allowlisted as long as they are not displayed in the UI so that the user can't change them. Would the user still be able to change them if they are just hidden?

clintandrewhall replied:
The third option would be to provide a list of settings from a package-- in this PR, `ALL_ALLOWLIST_SETTINGS`, and each solution sets the entire list one time from their setup contract.  I don't see a reason why this API can't be called multiple times, but I also see the value in calling it only once.

ElenaStoeva commented:
Thank you very much for the feedback @lukeelmers! It is really helpful.

I addressed your comments regarding moving the allowlist APIs to the server and left a couple of follow-up comments. Let me know what you think.

ElenaStoeva committed:
Don't allow updating a setting that is not allowlisted

ElenaStoeva commented on src/plugins/advanced_settings/server/plugin.ts:
```diff
@@ -21,6 +22,8 @@ export class AdvancedSettingsServerPlugin implements Plugin<object, object> {
 
     core.capabilities.registerProvider(capabilitiesProvider);
 
+    core.uiSettings.setAllowlist(ALL_ALLOWLIST_SETTINGS);
```
This is added here only for testing. It will be called from the serverless plugin instead once we register the Advanced settings app there.

clintandrewhall commented:
This is great @ElenaStoeva , and I think it's nearly shippable.  Once we answer a few more of these questions, we can send it to @elastic/kibana-core for feedback, (or send it now as a PRFC). :-)

ElenaStoeva committed:
Leverage readonly param and extract settings list into a package

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

ElenaStoeva commented on config/serverless.yml:
```diff
@@ -46,7 +46,7 @@ management.deeplinks.navLinkStatus: visible
 xpack.canvas.enabled: false
 xpack.cloud_integrations.data_migration.enabled: false
 data.search.sessions.enabled: false
-advanced_settings.enabled: false
```
This is only for testing the changes in serverless. It will not be merged.

ElenaStoeva commented on config/serverless.yml:
```diff
@@ -46,7 +46,7 @@ management.deeplinks.navLinkStatus: visible
 xpack.cloud_integrations.data_migration.enabled: false
 data.search.sessions.enabled: false
-advanced_settings.enabled: false
+advanced_settings.enabled: true
```
This is only for testing the changes in serverless. It will not be merged.

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

afharo commented:
I'd like to suggest an alternative:
We currently support `uiSettings.overrides`.

Could we leverage this and hide the YAML-overridden settings from the UI?

If something must have a value in Serverless, we can force-set it this way. WDYT?

clintandrewhall commented on packages/core/ui-settings/core-ui-settings-common/src/ui_settings.ts:
```diff
@@ -55,8 +55,12 @@ export interface UiSettingsParams<T = unknown> {
+   * 'strict' indicates that the value cannot be changed through API and is not displayed in the UI
+   * 'ui' indicates that the value is just not displayed in the UI
+   * */
+  readonly?: 'strict' | 'ui';
```
Unless I'm mistaken, this isn't a passive change.  If so, I wouldn't change this, but rather leave it as a `boolean`and add an internal API-- `setReadOnlyMode`-- that changes how a read-only field is _set_.

cc: @lukeelmers

ElenaStoeva replied:
I understand, I returned the `readonly` property to as it was and added a `readonlyMode` property instead. However, I'm not sure I understand correctly what `setReadOnlyMode` is supposed to do. Would this set the `readonlyMode` property to "strict"/"ui" when we're applying the allowlist, or is this supposed to be the setter that we would use specifically for updating `readonly` settings?

clintandrewhall commented on x-pack/plugins/serverless/server/plugin.ts:
```diff
@@ -75,6 +76,8 @@ export class ServerlessPlugin implements Plugin<ServerlessPluginSetup, Serverles
       );
     }
 
+    core.uiSettings.setAllowlist(ALL_ALLOWLIST_SETTINGS);
```
We need to expand this, and restrict `setAllowList` to be called one time.  We can also have `setAllowList` call `setReadOnlyMode('strict')`.

```
private setupProjectSettings(keys: string[] = []): void => {
  const settings = [...ALL_ALLOWLIST_SETTINGS].concat(keys);
  core.uiSettings.setAllowlist(settings);
}

// ...
public setup(core: CoreSetup) {
// ...
return {
  setupProjectSettings: this.setupProjectSettings
};
```
I can open a PR against this to demonstrate further, if necessary.

ElenaStoeva replied:
Thanks for the example, I added the `setupProjectSettings` from the serverless plugin and called it from the serverless project plugins as you suggested.

How would you suggest approaching restricting `setAllowlist` to be called only once? I was thinking about setting a private variable to false and toggling it when `setAllowlist` is called, but I remember you mentioned something about setting `eslint` rules. Can you share more about this?

clintandrewhall commented on packages/serverless/settings/README.mdx:
```diff
@@ -0,0 +1,10 @@
+date: 2023-08-24
+---
+
+This package contains common settings for Serverless projects.
```
nit: we should expand this with more detail.

ElenaStoeva replied:
Sure, I added some details. Let me know if there are any other specific details that you think might be helpful here.

ElenaStoeva commented:
> I'd like to suggest an alternative: We currently support `uiSettings.overrides`.
> 
> Could we leverage this and hide the YAML-overridden settings from the UI (AFAIK, the API won't allow overriding it)?
> 
> If something must have a value in Serverless, we can force-set it this way. WDYT?

Thanks for the suggestion, @afharo! Just a couple of questions to make sure I understand it correctly: 

Are all `uiSettings.overrides` YAML-overridden settings? If so, where are they being overridden?
Is what you are suggesting to override all settings that are not relevant in serverless so that they are not displayed in the UI and are not allowed to be updated though API?

clintandrewhall reviewed (commented):
This is a great improvement.  There's a few minor changes to the API that I think will make it more obvious how the `allowList` is constructed, and make it more secure.

clintandrewhall commented:
> Could we leverage this and hide the YAML-overridden settings from the UI

@afharo The list of settings available in Serverless is _significantly_ shorter.  We're allow-listing those on a project-by-project basis, and a YML-based approach wouldn't be practical.  

In summary: we're not  _changing_ the settings, we're shortening the list available to API and UI changes.

afharo commented:
> Are all `uiSettings.overrides` YAML-overridden settings? If so, where are they being overridden?

Yes, here's an example of a UI Setting override: https://github.com/elastic/kibana/blob/5136714f6fcd0992c275cd63b9b72bcb9d285008/config/serverless.security.yml#L18

At the moment, we show it in the UI but it cannot be edited:
<img width="2249" alt="image" src="https://github.com/elastic/kibana/assets/5469006/faf8c611-3a5d-4886-bfe3-ee5e507bb066">

We could set the defaults in `serverless.yml`, and per-project in `serverless.{project}.yml`.

> Is what you are suggesting to override all settings that are not relevant in serverless so that they are not displayed in the UI and are not allowed to be updated through API?

I guess I am. I understand this may be undesirable because the list of settings to hide is larger than the ones to show.
However, I wonder if it's worth triggering the discussion about the defaults that make sense for each one. After all, Serverless is a different offering and things like `timepicker:refreshIntervalDefaults` might need to change.

I also understand that flipping it from allowlist to _hide_ list (via overrides) might lead to the setting being exposed even when it shouldn't, if we fail to review that during PRs. While allowlist makes it more obvious (_it doesn't show up... why? oh... right!_).

If we are going the allowlist... should we maintain an on-premise allowlist as well? There may be settings that are intended only for serverless, won't they?

lukeelmers commented:
> I wonder if it's worth triggering the discussion about the defaults that make sense for each one.

There is some discussion happening in this area already around limits for some of these settings, especially with security solution. There's an internal document I can link you to. 

> flipping it from allowlist to hide list (via overrides) might lead to the setting being exposed even when it shouldn't, if we fail to review that during PRs. While allowlist makes it more obvious (it doesn't show up... why? oh... right!).

This was the motivation for not using `uiSettings.overrides`. The concern is a scenario where a user is able to circumvent a setting that's readonly in the UI by hitting the server endpoint directly.

> If we are going the allowlist... should we maintain an on-premise allowlist as well? There may be settings that are intended only for serverless, won't they?

I think this is less of a problem for now, because settings intended only for serverless should only be registered in a serverless environment (e.g. in one of the `serverless*` plugins) and therefore wouldn't show up in self-managed.

My ideal future state would be that `readonly` on a setting behaves as you'd expect in both serverless & self-managed -- that is, it enforces readonly not just on the client as it does today, but also on the server. But because you can technically circumvent readonly settings by hitting the API directly today, there's a risk that this is a breaking change for self-managed users who may have some sort of tooling in place to change readonly settings by API. So this is something we'd need to save for a 9.0 stack release.

afharo commented:
> There is some discussion happening in this area already around limits for some of these settings, especially with security solution. There's an internal document I can link you to.

Nice! Thanks for confirming! I hope we can clean up any of the deprecated/unused ones 😇 

> This was the motivation for not using `uiSettings.overrides`. The concern is a scenario where a user is able to circumvent a setting that's readonly in the UI by hitting the server endpoint directly.

I just tested it and this is not possible today. Trying to override the `defaultRoute` via API `POST /internal/kibana/settings` fails with `400`: 
```JSON
{"statusCode":400,"error":"Bad Request","message":"Unable to update \"defaultRoute\" because it is overridden"}
```

I even tried exporting the SO, tampering, and importing it. While stored in the SO, the YAML override takes precedence. If we follow the allowlist approach, we need to make sure that we dismiss any user values.

My concern is that we may forget to set an override to a setting that should not be visible in Advanced Settings because it's _opt-out_ instead of being _opt-in_.

> I think this is less of a problem for now, because settings intended only for serverless should only be registered in a serverless environment (e.g. in one of the `serverless*` plugins) and therefore wouldn't show up in self-managed.

I think this goes against the recommendation (https://github.com/elastic/kibana/issues/101907).

If we are looking for explicit opt-ins, I'd try to design an additional property in the UI Settings register API that would define in which offering it should be visible. This allowlist would require developers to change _yet another unrelated_ plugin whenever they add a new setting.

Having said that, I won't oppose shipping this now and revisiting it in the future (given the times, that it's serverless, hence internal, and the low amount of affected settings).

clintandrewhall commented:
> If we are going the allowlist... should we maintain an on-premise allowlist as well? There may be settings that are intended only for serverless, won't they?

This is a very good point... but we don't have those today, AFAIK.  But I don't think we would register those settings in Kibana, but rather in the `serverless` plugin, correct?  In other words, we'd be _adding_ settings in Serverless, rather than making common settings `readonly` in Serverless... so it's a different use case.

Perhaps a `denyList` would be a good addition if that case arises-- where we add a Serverless setting to "classic" Kibana.  But I think it's out of scope for now.

ElenaStoeva committed:
Expose setupProjectSettings api from serverless plugin

lukeelmers commented:
> My concern is that we may forget to set an override to a setting that should not be visible in Advanced Settings because it's opt-out instead of being opt-in.

Same! uiSettings.overrides gives us all of the protections we want _except_ for this. So the question this PR is trying to solve is how to provide this same type of enforcement in an opt-in manner.

>> I think this is less of a problem for now, because settings intended only for serverless should only be registered in a serverless environment (e.g. in one of the serverless* plugins) and therefore wouldn't show up in self-managed.
>
> I think this goes against the recommendation (https://github.com/elastic/kibana/issues/101907).

Not necessarily, my point (and I think what Clint was saying above) is that such configs could be registered from the solution-specific serverless plugin, which is where they are ideally also being consumed (e.g. `serverless_observability` plugin would register serverless-only settings and also consume them in the serverless-only UI they have constructed).

ElenaStoeva committed:
Change readonly param and improve settings package readme

ElenaStoeva committed:
Extract search and security settings into packages

ElenaStoeva committed:
Fix types errors

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

ElenaStoeva committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

ElenaStoeva committed:
Fix event log tests

ElenaStoeva commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

ElenaStoeva committed:
Fix alerting test mocks

ElenaStoeva committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

pgayvallet commented on packages/serverless/settings/common/index.ts:
```diff
@@ -0,0 +1,59 @@
+ * Side Public License, v 1.
+ */
+
+const GENERAL_SETTINGS = [
+  'csv:quoteValues',
+  'dateFormat:dow',
+  'dateFormat:scaled',
+  'dateFormat:tz',
```
It's probably a NIT, but I'm somewhat concerned about having this list of setting IDs being fully decoupled from the actual setting definitions. 

I'm a bit scared of human error coming from developers not knowing about this list when they, say, add a new setting, or remove an existing one.

Do we have anything in mind to avoid the risk of having this list "desync" from the actual definitions?

clintandrewhall replied:
I'd like this to be in a package, rather than coded as strings.

pgayvallet replied:
We could have the ids of the settings in a package, and then imported/used both by the allowlists and by the actual settings definitions.

Or maybe the setting definitions themselves could be moved into packages and then the allowlist importing the ids from here?

Either way it's a lot of "moving code", and not a blocker for the initial implementation

ElenaStoeva replied:
Both approaches make sense to me. I think the second one would require significantly more "moving code" changes so we might need to defer it to a separate PR, but the first approach could be implemented as part of this PR - or at least for now we can make packages for the ids and use them in the allowlists, and in a later PR to import them in all places with definitions where the settings are registered.

pgayvallet commented on x-pack/plugins/serverless/server/plugin.ts:
```diff
@@ -36,6 +37,11 @@ const typeToIdMap: Record<ProjectType, string> = {
 export class ServerlessPlugin implements Plugin<ServerlessPluginSetup, ServerlessPluginStart> {
   private readonly config: ServerlessConfig;
 
+  private setupProjectSettings(core: CoreSetup, keys: string[]): void {
+    const settings = [...ALL_COMMON_SETTINGS].concat(keys);
+    core.uiSettings.setAllowlist(settings);
+  }
```
- if the plugin receives multiple calls to `setupProjectSettings`, then only the value from the last one will effectively be used.
- if the plugin receives **no** call to `setupProjectSettings`, then `core.uiSettings.setAllowlist` will never be called, meaning that all settings will be enabled / visible.

Are we fine with this? I know that "in theory", exactly one call to `setupProjectSettings` will be done for any project type in serverless, but how much do we want to rely blindly on this "theory" (e.g we add a 4rth project type later, by default all settings will be enabled and visible)?

pgayvallet replied:
cc @jloleysens FYI

Yeah, that's not the first time we're facing problems with this script that forces all plugins to be enabled at the same time (which is something that was fine when we designed the thing but no longer with the "project type only" plugins.

I don't have any great suggestions right now, so I would just disable this check for now to avoid blocking the PR, and try to address it as a follow up.

pgayvallet reviewed (commented):
I think I'm fine with the approach.

I have a few questions and remarks though:

ElenaStoeva committed:
Address CR comments

ElenaStoeva commented on packages/core/ui-settings/core-ui-settings-common/src/ui_settings.ts:
```diff
@@ -57,6 +57,12 @@ export interface UiSettingsParams<T = unknown> {
   requiresPageReload?: boolean;
   /** a flag indicating that value cannot be changed */
   readonly?: boolean;
+  /**
+   * a flag indicating the level of restriction of the readonly settings
+   * 'strict' indicates that the value cannot be changed through API and is not displayed in the UI
+   * 'ui' indicates that the value is just not displayed in the UI
+   * */
+  readonlyMode?: 'strict' | 'ui';
```
Currently, settings with `readonly: true` are [filtered out from the UI](https://github.com/elastic/kibana/blob/4b02740b32542b6e84ee7a8abec9cb355f70db35/src/plugins/advanced_settings/public/management_app/settings_helper.ts#L26), but there is no enforcement for prohibiting `set` on these settings. Since we want to prohibit `set` for non-allowlisted settings, the idea is to add two readonly modes - `ui`, which basically has the current behaviour of `readonly: true`, and `strict`, which also restricts `set`. We want to add these modes so that these changes don't affect the current behaviour of any of the existing settings that are readonly.

We started discussing this approach with @clintandrewhall and @lukeelmers last week (see https://github.com/elastic/kibana/pull/164471#discussion_r1304396905 and https://github.com/elastic/kibana/pull/164471#discussion_r1305667169) but we haven't clarified some details yet - we have a sync-up today so we will discuss this further.

ElenaStoeva commented on packages/core/ui-settings/core-ui-settings-server-internal/src/ui_settings_service.ts:
```diff
@@ -148,6 +153,25 @@ export class UiSettingsService
+      }
+    }
+    for (const [key, definition] of this.uiSettingsGlobalDefaults) {
+      if (!keys.includes(key)) {
```
Good catch! I added this with [b7a8350](https://github.com/elastic/kibana/pull/164471/commits/b7a83504f1f7d47d28c9fc262977042548a96c2d)

ElenaStoeva committed:
Add setReadonlyMode method and enforcements/validations

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

ElenaStoeva committed:
Add tests for setAllowlist and validateAllowlist

ElenaStoeva committed:
Disable mappings check

ElenaStoeva committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

ElenaStoeva committed:
Extract setting ids into a package

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

ElenaStoeva committed:
Fix logic, refactor changes

ElenaStoeva committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

ElenaStoeva committed:
Fix actions plugin tests

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

ElenaStoeva committed:
Skip blocking test

ElenaStoeva committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

ElenaStoeva committed:
Add tests for applyAllowlist

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

ElenaStoeva committed:
Add more documentation

ElenaStoeva committed:
Add serverless plugin test

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

pmuellr commented on x-pack/plugins/alerting/server/plugin.test.ts:
```diff
@@ -242,6 +243,7 @@ describe('Alerting Plugin', () => {
               features: featuresPluginMock.createSetup(),
               unifiedSearch: autocompletePluginMock.createSetupContract(),
               ...(useDataStreamForAlerts ? { serverless: {} } : {}),
+              serverless: serverlessPluginMock.createSetupContract(),
```
I don't think this is going to work right, or is at least confusing.  The new line sets a `serverless` property, but so does the line above, optionally.  In addition, the line above only sets the value when `useDataStreamForAlerts` is true.  With the new code, it will always be set, unless for some reason the optional one above it "wins" (which I don't think it will).  

I was surprised no tests failed because of this, but this is testing the plugin `start()` method, which doesn't currently have dependencies on the `serverless` plugin.  If it did, I'd guess we'd get some failures.

The intention here is that we are testing "with" and "without" the serverless plugin being available, which is determined in the tests via the `useDataStreamForAlerts` value.  We are basically not testing the case of the `serverless` plugin **NOT** being available, only testing it **BEING** available.

I think the code should be changed to the version on line 76 now.  

Same is true for the similiar changes below ...

pmuellr commented on x-pack/plugins/event_log/server/plugin.test.ts:
```diff
@@ -42,7 +44,7 @@ describe('event_log plugin', () => {
     const spaces = spacesMock.createStart();
     // serverless setup is currently empty, and there is no mock
-    plugin.setup(coreSetup, { serverless: {} });
+    plugin.setup(coreSetup, { serverless: { setupProjectSettings(keys: string[]) {} } });
```
Seems like we should use the mock here, so we don't have to change this code when the interface changes.

mistic committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

pmuellr commented on x-pack/plugins/alerting/server/plugin.test.ts:
```diff
@@ -74,7 +75,7 @@ describe('Alerting Plugin', () => {
           unifiedSearch: autocompletePluginMock.createSetupContract(),
           // serverless setup is currently empty, and there is no mock
-          ...(useDataStreamForAlerts ? { serverless: {} } : {}),
+          ...(useDataStreamForAlerts ? { serverless: { setupProjectSettings: jest.fn() } } : {}),
```
Can we use the mock here instead of creating the mocks fn's manually?

pmuellr reviewed (changes requested):
I pointed out some code that should be changed in the alerting plugin regarding setting the serverless plugin mock (or not).  I tried to provide some explanation, but it may still not be obvious what's going on, let me know if that doesn't make sense.

Also a few spots that I think we can use a Mock for, instead of "hand-building" the mock objects.

ElenaStoeva committed:
Fix alerting and event_log tests

ElenaStoeva commented:
> I pointed out some code that should be changed in the alerting plugin regarding setting the serverless plugin mock (or not). I tried to provide some explanation, but it may still not be obvious what's going on, let me know if that doesn't make sense.
> 
> Also a few spots that I think we can use a Mock for, instead of "hand-building" the mock objects.

Thanks for pointing this out @pmuellr! Not sure why I mistakenly added the redundant serverless dependency and the "hand-built" mock objects. Please let me know if it is correct now.

pmuellr commented on x-pack/plugins/alerting/server/plugin.test.ts:
```diff
@@ -74,7 +75,9 @@ describe('Alerting Plugin', () => {
           features: featuresPluginMock.createSetup(),
           unifiedSearch: autocompletePluginMock.createSetupContract(),
           // serverless setup is currently empty, and there is no mock
```
1 / 3 - Could you remove this comment since it's obviously no longer correct :-)

pmuellr commented on x-pack/plugins/event_log/server/plugin.test.ts:
```diff
@@ -42,7 +45,7 @@ describe('event_log plugin', () => {
     const plugin = new Plugin(initializerContext);
     const spaces = spacesMock.createStart();
     // serverless setup is currently empty, and there is no mock
```
3 / 3 - Could you remove this comment since it's obviously no longer correct :-)

pmuellr commented on x-pack/plugins/event_log/server/plugin.test.ts:
```diff
@@ -19,7 +20,9 @@ describe('event_log plugin', () => {
 
     const plugin = new Plugin(initializerContext);
     // serverless setup is currently empty, and there is no mock
```
2 / 3 - Could you remove this comment since it's obviously no longer correct :-)

pmuellr reviewed (approved):
code LGTM - thanks for the requested changes

If you have time, could you remove the 3 comments I pointed out, which are now incorrect? :-). I knew I never should have added those!

ElenaStoeva committed:
Remove unnecessary comments

afharo commented on packages/core/ui-settings/core-ui-settings-common/src/ui_settings.ts:
```diff
@@ -25,6 +25,14 @@ export type UiSettingsType =
+ * 'ui' indicates that the value is just not displayed in the UI
+ * @public
+ * */
+export type ReadonlyModeType = 'strict' | 'ui';
```
please, remember to export this in `index.ts`

afharo commented on .buildkite/scripts/steps/checks.sh:
```diff
@@ -21,4 +21,5 @@ export DISABLE_BOOTSTRAP_VALIDATION=false
 .buildkite/scripts/steps/checks/test_hardening.sh
 .buildkite/scripts/steps/checks/ftr_configs.sh
-.buildkite/scripts/steps/checks/saved_objects_compat_changes.sh
+# Skipped as blocking https://github.com/elastic/kibana/pull/164471
+#.buildkite/scripts/steps/checks/saved_objects_compat_changes.sh
```
I don't feel comfortable skipping this check. It's critical to keep smooth upgrades.

ElenaStoeva replied:
This one was failing because it enables all plugins [here](https://github.com/elastic/kibana/blob/1c7ac5d5de0f7f14c0e23bc87bd89119b5a126df/packages/kbn-check-mappings-update-cli/src/extract_mappings_from_plugins.ts#L35), so multiple serverless project plugins call `setupProejctSettings`, which throws the error `The uiSettings allowlist has already been set up. ...`.

Do you have any suggestion for how we can fix this test? I'm not really familiar with it, but I imagine we could not enable the serverless project plugins (or just enable one), but I'm not sure if that would be a good idea.

pgayvallet replied:
I thought we agreed on disabling the multi/no call check (at least for now) given https://github.com/elastic/kibana/pull/164471#discussion_r1314478064?

ElenaStoeva replied:
> 



> I thought we agreed on disabling the multi/no call check (at least for now) given [#164471 (comment)](https://github.com/elastic/kibana/pull/164471#discussion_r1314478064)?

Sorry, it seems I misunderstood what you meant to disable. So instead of disabling the failing tests, we don't throw an error (for now) when `setAllowlist()` is called multiple times, correct?
I think the no-call check doesn't cause any failures.

afharo commented on src/core/server/integration_tests/config/check_dynamic_config.test.ts:
```diff
@@ -64,7 +64,8 @@ describe('checking migration metadata changes on all registered SO types', () =>
    * The intent is to trigger a code review from the Core and Security teams to discuss potential issues.
    */
-  test('detecting all the settings that have opted-in for dynamic in-memory updates', () => {
+  // Skipped as blocking https://github.com/elastic/kibana/pull/164471
+  test.skip('detecting all the settings that have opted-in for dynamic in-memory updates', () => {
```
Can we address the reason for skipping the test instead?

ElenaStoeva replied:
Haven't looked into this test in depth, but it seems its `beforeAll` hook is failing with the same error - `The uiSettings allowlist has already been set up. ...`, which makes me think that it enables multiple serverless project plugins. Do you know if that's the case?

I will look into it further to see what could be done.

ElenaStoeva replied:
From what I see, it looks this test indeed enables all plugins [here](https://github.com/elastic/kibana/blob/b335c3abbcd776a5d9d6b4e042e2165478cc42ae/packages/core/root/core-root-server-internal/src/server.ts#L349), which is why we get this error thrown.

afharo commented on x-pack/plugins/serverless_search/server/types.ts:
```diff
@@ -8,6 +8,7 @@
 import type { SecurityPluginStart } from '@kbn/security-plugin/server';
 import type { EnterpriseSearchPluginStart } from '@kbn/enterprise-search-plugin/server';
 import type { MlPluginSetup } from '@kbn/ml-plugin/server';
+import { ServerlessPluginSetup } from '@kbn/serverless/server';
```
nit: `import type { ...` 😇

ElenaStoeva replied:
Good catch! 😄

afharo reviewed (changes requested):
Overall, LGTM! Great job!

However, I don't feel comfortable skipping the SO compat changes check. It's critical to keep smooth upgrades. Can we reenable and address the reasons for the failures?

As a side comment, can we add FTRs in `test_serverless` 😇 ?

ElenaStoeva committed:
Fix exporting/importing errors

thomheymann reviewed (approved):
security changes LGTM

ElenaStoeva committed:
Remove old security setting id

ElenaStoeva committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

machadoum reviewed (approved):
Security Solutions changes LGTM!

@ElenaStoeva While I was testing how the advanced settings page looks on the security solution serverless I noticed that the tabs don't make much sense. 

<img width="1540" alt="Screenshot 2023-09-08 at 14 00 44" src="https://github.com/elastic/kibana/assets/1490444/a984b28f-1d3f-4a0a-b641-6c60c96e2356">

Serverless doesn't support multiple spaces and apparently has no global setting. So we could get rid of the tabs. 


This isn't a blocker for your PR, but we need to address it to enable the page on serverless.

ElenaStoeva committed:
Add functional tests

ElenaStoeva committed:
Merge branch 'advanced_settings/allowlist_for_serverless' of https://github.com/ElenaStoeva/kibana into advanced_settings/allowlist_for_serverless

ElenaStoeva commented:
Thanks for the review @machadoum!

This PR is only for the changes on the uiSettings service. I enabled the existing Advanced settings plugin just so we can test these changes in the serverless UI, but I will disable it before I merge. We will make a separate PR for the serverless Advanced settings UI, which will not include tabs for Space/Global settings.

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

ElenaStoeva committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

ElenaStoeva committed:
Unskip failed tests and disable multi-call check for setAllowlist

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

ElenaStoeva committed:
Skip test for setAllowlist check

afharo commented on x-pack/test_serverless/functional/test_suites/observability/advanced_settings.ts:
```diff
@@ -0,0 +1,42 @@
+      expect(url).to.contain(`/settings`);
+    });
+
+    describe('renders common settings', () => {
```
nit: `renders observability settings` 😇

afharo reviewed (approved):
LGTM! Thank you for iterating and pushing it through the finish line!

ElenaStoeva committed:
Fix test descriptions

dmlemeshko reviewed (approved):
Appex-QA changes LGTM

Dosant commented on packages/core/ui-settings/core-ui-settings-server-internal/src/ui_settings_service.ts:
```diff
@@ -148,6 +159,55 @@ export class UiSettingsService
+    this.allowlist?.forEach((key) => {
+      if (!this.uiSettingsDefaults.has(key) && !this.uiSettingsGlobalDefaults.has(key)) {
+        throw new Error(
+          `The uiSetting with key [${key}] is in the allowlist but is not registered. Make sure to remove it from the allowlist in /packages/serverless/settings`
```
I was wondering if this is too strict. The default allow list is in a common package but the settings are registered by plugins. I wonder if any of those plugins can be disabled and then this throughs.

afharo replied:
This is a good point... We may want to apply the validation in dev mode only.

While `serverless` should be served as-is (no custom enabling/disabling of plugins per customer), and disabling plugins should be our last resort, as it could lead to unexpected and untested behavior, I wonder if we want to allow disabling plugins when there's a specific issue in prod that we want to mitigate.

cc @lukeelmers

ElenaStoeva replied:
> This is a good point... We may want to apply the validation in dev mode only.

Sure, this makes sense to me. I added this condition with https://github.com/elastic/kibana/pull/164471/commits/6bfe3da1a07327c64d97ed9d7f6e6e4029efc62a

Dosant reviewed (approved):
I tested and reviewed the code. lgtm. nice additions of serverless tests 👍

ElenaStoeva committed:
Validate allowlist only in development mode

ElenaStoeva committed:
Merge branch 'main' into advanced_settings/allowlist_for_serverless

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

lukeelmers commented on packages/core/ui-settings/core-ui-settings-server-internal/src/ui_settings_service.ts:
```diff
@@ -94,10 +99,19 @@ export class UiSettingsService
 
   public async start(): Promise<InternalUiSettingsServiceStart> {
+    if (this.allowlist) {
+      // If we are in development mode, check if all settings in the allowlist are registered
+      if (process.env.NODE_ENV !== 'production') {
+        this.validateAllowlist();
+      }
```
You shouldn't need to read from `process.env` directly here -- we expose a proper mechanism for determining dev mode from the `Env` service in core. In this service you should be able to access it from `coreContext`:

```js
class UiSettingsService {
  constructor(coreContext) {
    this.isDev = coreContext.env.mode.dev;
  }

  async start() {
    if (this.allowlist) {
      if (this.isDev) {
        this.validateAllowlist();
      }
      // ...etc
    }
  }
}
```

ElenaStoeva replied:
Got it, thanks for letting me know!

ElenaStoeva committed:
Refactor isDev check

saarikabhasi reviewed (approved):
Noticed no global settings similar this [comment](https://github.com/elastic/kibana/pull/164471#pullrequestreview-1617347056) during testing and as it would be [removed](https://github.com/elastic/kibana/pull/164471#issuecomment-1711586752), should be fine I guess.   Search project code changes LGTM !

ElenaStoeva committed:
Disable Advanced settings app in serverless and skip functional tests
